common/src/chain/transaction/output/stakelock.rs

```diff
@@ -33,13 +34,15 @@ pub struct StakePoolData {
 
 impl StakePoolData {
     pub fn new(
+        value: Amount,
```

Comment:
Maybe there should be a type distinction between Mlt coins and other token amounts. There is already a `Mlt` type somewhere in chain config. We can then distinguish which output types accept arbitrary tokens vs Mlts only more granulary.

Reply:
This is exactly what's done in this PR. For staking, you can only stake MLT. Hence, it's an amount. For other outputs, it's OutputValue enum, so can be anything.

Reply:
Well `Amount` is also used in `OutputValue` to represent token amount (see `TokenTransfer`), so the separation is not as tight as it could be

Reply:
This is, again, the same issue we're having with timestamps, where they're also just numbers that can't be interpreted without assumptions when coming from bytes. Unfortunately, it's an unsolvable problem. A number is just a number as long as it touches bytes. Without context, we can't deserialize it into a type without making an assumption. This is why an `Amount` is just an amount until it gets into an output. Only then, we can make an assumption on what that `Amount` really means; i.e., whether it's a coin or a token.

I'm not saying it's impossible to find a trick to do all this... but it's, at least, not so obvious that it can be done given the current constraints.

Reply:
To be clear about the `Mlt` type, it's not a good choice to involve the blockchain operation with how many decimals we're using. This is the job of the wallets and user-touching stuff. It doesn't really add value unless we involve the context of ChainConfig.

Reply:
You can't really benefit much from this, and here's why: Amount, itself, is already type-safe (and cannot be added to timestamps, for example). However, we can have any number of tokens defined at runtime, and they're not distinguishable at compile-time. So, the best we can do is distinguish between ML and other tokens. But tokens among each other will still have one amount type at best.

Alrighty. How about adding a generic parameter to Amount to make that distinction? We can make a default generic type, and specialize it for Amount where appropriate. Sounds good? This will have the added benefit that all math ops will remain valid, instead of having to reintroduce them, which to be honest is a big problem with extra types on things that support math.

Reply:
> Alrighty. How about adding a generic parameter to Amount to make that distinction? We can make a default generic type tag, and specialize it for ML where appropriate. Sounds good? This will have the added benefit that all math ops will remain valid, instead of having to reintroduce them, which to be honest is a big problem with extra types on things that support math.

That sounds reasonable.

Regarding tokens not being distinguishible from each other at compile time, I think it can be partially mitigated by a type that enforces the check at runtime. I.e. instead of having plain amount, you have amount + token ID. Arithmetic operations and similar check you are adding same kind of tokens, and return a `Result` as appropriate. Also such type would probably implement `PartialEq`, `PartialOrd` but not `Eq`, `Ord`.